File: packages/sdk-python/rollgate/events.py
"""
Event collector for A/B testing conversion tracking.
"""

import asyncio
import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional, Dict, Any, List

import httpx

logger = logging.getLogger("rollgate.events")


@dataclass
class TrackEventOptions:
    """Options for tracking a conversion event."""

    flag_key: str
    event_name: str
    user_id: str
    variation_id: Optional[str] = None
    value: Optional[float] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class EventCollectorConfig:
    """Configuration for the event collector."""

    flush_interval_ms: int = 30000
    max_buffer_size: int = 100
    enabled: bool = True


class EventCollector:
    """
    Buffers and batches conversion events for A/B testing.

    Events are buffered in memory and flushed periodically or when
    the buffer reaches max_buffer_size.
    """

    def __init__(
        self,
        endpoint: str,
        api_key: str,
        config: EventCollectorConfig,
        http_client: httpx.AsyncClient,
    ):
        self._endpoint = endpoint
        self._api_key = api_key
        self._config = config
        self._http_client = http_client
        self._buffer: List[Dict[str, Any]] = []
        self._flush_task: Optional[asyncio.Task] = None
        self._closing = False

    def start(self) -> None:
        """Start the periodic flush task."""
        if not self._config.enabled:
            return
        self._flush_task = asyncio.create_task(self._flush_loop())

    async def stop(self) -> None:
        """Stop the collector and flush remaining events."""
        self._closing = True
        if self._flush_task:
            self._flush_task.cancel()
            try:
                await self._flush_task
            except asyncio.CancelledError:
                pass
        # Best-effort final flush
        try:
            await self.flush()
        except Exception:
            pass

    def track(self, options: TrackEventOptions) -> None:
        """Add an event to the buffer."""
        if not self._config.enabled:
            return

        event: Dict[str, Any] = {
            "flagKey": options.flag_key,
            "eventName": options.event_name,
            "userId": options.user_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }

        if options.variation_id is not None:
            event["variationId"] = options.variation_id
        if options.value is not None:
            event["value"] = options.value
        if options.metadata is not None:
            event["metadata"] = options.metadata

        self._buffer.append(event)

        if len(self._buffer) >= self._config.max_buffer_size:
            asyncio.create_task(self._flush_quiet())

    async def flush(self) -> None:
        """Flush all buffered events to the server."""
        if not self._buffer:
            return

        events = self._buffer.copy()
        self._buffer.clear()

        try:
            response = await self._http_client.post(
                self._endpoint,
                json={"events": events},
                headers={
                    "Authorization": f"Bearer {self._api_key}",
                    "Content-Type": "application/json",
                },
            )

            if response.status_code >= 400:
                # Re-buffer on failure
                self._re_buffer(events)
                logger.warning(f"Event flush failed with status {response.status_code}")

        except Exception as e:
            self._re_buffer(events)
            logger.warning(f"Event flush error: {e}")

    @property
    def buffer_size(self) -> int:
        """Get the current number of buffered events."""
        return len(self._buffer)

    def _re_buffer(self, events: List[Dict[str, Any]]) -> None:
        """Re-add events to buffer on failure."""
        combined = events + self._buffer
        max_size = self._config.max_buffer_size * 2
        if len(combined) > max_size:
            combined = combined[-max_size:]
        self._buffer = combined

    async def _flush_loop(self) -> None:
        """Periodically flush events."""
        interval = self._config.flush_interval_ms / 1000
        while not self._closing:
            await asyncio.sleep(interval)
            if self._closing:
                break
            await self._flush_quiet()

    async def _flush_quiet(self) -> None:
        """Flush without raising."""
        try:
            await self.flush()
        except Exception:
            pass

File: packages/sdk-flutter/lib/src/events.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Options for tracking a conversion event.
class TrackEventOptions {
  final String flagKey;
  final String eventName;
  final String userId;
  final String? variationId;
  final double? value;
  final Map<String, dynamic>? metadata;

  TrackEventOptions({
    required this.flagKey,
    required this.eventName,
    required this.userId,
    this.variationId,
    this.value,
    this.metadata,
  });
}

/// Configuration for the event collector.
class EventCollectorConfig {
  final int flushIntervalMs;
  final int maxBufferSize;
  final bool enabled;

  const EventCollectorConfig({
    this.flushIntervalMs = 30000,
    this.maxBufferSize = 100,
    this.enabled = true,
  });
}

/// Buffers and batches conversion events for A/B testing.
class EventCollector {
  final String _endpoint;
  final String _apiKey;
  final EventCollectorConfig _config;
  final http.Client _httpClient;

  final List<Map<String, dynamic>> _buffer = [];
  Timer? _flushTimer;

  EventCollector({
    required String endpoint,
    required String apiKey,
    EventCollectorConfig config = const EventCollectorConfig(),
    required http.Client httpClient,
  })  : _endpoint = endpoint,
        _apiKey = apiKey,
        _config = config,
        _httpClient = httpClient;

  /// Start the periodic flush timer.
  void start() {
    if (!_config.enabled) return;
    _flushTimer = Timer.periodic(
      Duration(milliseconds: _config.flushIntervalMs),
      (_) => flush(),
    );
  }

  /// Stop the collector and flush remaining events.
  Future<void> stop() async {
    _flushTimer?.cancel();
    _flushTimer = null;
    try {
      await flush();
    } catch (_) {}
  }

  /// Track a conversion event.
  void track(TrackEventOptions options) {
    if (!_config.enabled) return;

    final event = <String, dynamic>{
      'flagKey': options.flagKey,
      'eventName': options.eventName,
      'userId': options.userId,
      'timestamp': DateTime.now().toUtc().toIso8601String(),
    };

    if (options.variationId != null) event['variationId'] = options.variationId;
    if (options.value != null) event['value'] = options.value;
    if (options.metadata != null) event['metadata'] = options.metadata;

    _buffer.add(event);

    if (_buffer.length >= _config.maxBufferSize) {
      flush();
    }
  }

  /// Flush all buffered events to the server.
  Future<void> flush() async {
    if (_buffer.isEmpty) return;

    final events = List<Map<String, dynamic>>.from(_buffer);
    _buffer.clear();

    try {
      final response = await _httpClient
          .post(
            Uri.parse(_endpoint),
            headers: {
              'Authorization': 'Bearer $_apiKey',
              'Content-Type': 'application/json',
            },
            body: jsonEncode({'events': events}),
          )
          .timeout(const Duration(seconds: 10));

      if (response.statusCode >= 400) {
        _reBuffer(events);
      }
    } catch (_) {
      _reBuffer(events);
    }
  }

  /// Get the current buffer size.
  int get bufferSize => _buffer.length;

  void _reBuffer(List<Map<String, dynamic>> events) {
    final combined = [...events, ..._buffer];
    final maxSize = _config.maxBufferSize * 2;
    _buffer.clear();
    if (combined.length > maxSize) {
      _buffer.addAll(combined.sublist(combined.length - maxSize));
    } else {
      _buffer.addAll(combined);
    }
  }
}
